addons/controllers/tanzuclusterbootstrap_controller.go

```diff
@@ -0,0 +1,607 @@
+
+	dataValuesSecretMutateFn := func() error {
+		dataValuesSecret.Data = map[string][]byte{}
+		for k, v := range secret.Data {
+			dataValuesSecret.Data[k] = v
+		}
+		return nil
+	}
+
+	_, err := controllerutil.CreateOrPatch(r.context, clusterClient, dataValuesSecret, dataValuesSecretMutateFn)
+	if err != nil {
+		log.Error(err, "Error creating or patching addon data values secret")
+		return nil, err
+	}
+	return dataValuesSecret, nil
+}
+
+//nolint:gocyclo //gocyclo complains complexity of 18 compared to 15, this function has been refactored to maximum to satisfy gocyclo
+// cloneSecretsAndProviders clones linked secrets and providers into the same namespace as TanzuClusterBootstrap
+func (r *tanzuClusterBootstrapReconciler) cloneSecretsAndProviders(cluster *clusterapiv1beta1.Cluster, bootstrap *runtanzuv1alpha3.TanzuClusterBootstrap,
+	templateNS string, log logr.Logger) ([]*corev1.Secret, []*unstructured.Unstructured, error) {
+
+	var createdProviders []*unstructured.Unstructured
+	var createdSecrets []*corev1.Secret
+
+	aggregateSecretsProviders := func(cluster *clusterapiv1beta1.Cluster, bootstrap *runtanzuv1alpha3.TanzuClusterBootstrap,
+		valuesFrom *runtanzuv1alpha3.ValuesFrom, templateNS string, packageShortName string, log logr.Logger) error {
+		secret, provider, err := r.updateValues(cluster, bootstrap, valuesFrom, templateNS, packageShortName, log)
+		if err != nil {
+			return err
+		}
+		if secret != nil {
+			createdSecrets = append(createdSecrets, secret)
+		}
+		if provider != nil {
+			createdProviders = append(createdProviders, provider)
+		}
+		return nil
+	}
+
+	if bootstrap.Spec.CNI != nil && bootstrap.Spec.CNI.ValuesFrom != nil {
+		if err := aggregateSecretsProviders(cluster, bootstrap, bootstrap.Spec.CNI.ValuesFrom, templateNS, packageShortName(bootstrap.Spec.CNI.RefName), log); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if bootstrap.Spec.CPI != nil && bootstrap.Spec.CPI.ValuesFrom != nil {
+		if err := aggregateSecretsProviders(cluster, bootstrap, bootstrap.Spec.CPI.ValuesFrom, templateNS, packageShortName(bootstrap.Spec.CPI.RefName), log); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if bootstrap.Spec.CSI != nil && bootstrap.Spec.CSI.ValuesFrom != nil {
+		if err := aggregateSecretsProviders(cluster, bootstrap, bootstrap.Spec.CSI.ValuesFrom, templateNS, packageShortName(bootstrap.Spec.CSI.RefName), log); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if bootstrap.Spec.Kapp != nil && bootstrap.Spec.Kapp.ValuesFrom != nil {
+		if err := aggregateSecretsProviders(cluster, bootstrap, bootstrap.Spec.Kapp.ValuesFrom, templateNS, packageShortName(bootstrap.Spec.Kapp.RefName), log); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	for _, additionalPackage := range bootstrap.Spec.AdditionalPackages {
+		if err := aggregateSecretsProviders(cluster, bootstrap, additionalPackage.ValuesFrom, templateNS, packageShortName(additionalPackage.RefName), log); err != nil {
```

Comment:
Potential nil pointer dereference here: additionalPackage is a pointer.

Reply:
not an issue anymore since this is refactored